src/Servers/Kestrel/Core/src/Middleware/TlsListenerMiddleware.cs

```diff
@@ -0,0 +1,125 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using Microsoft.AspNetCore.Connections;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Middleware;
+
+internal sealed class TlsListenerMiddleware
+{
+    private readonly ConnectionDelegate _next;
+    private readonly Action<ConnectionContext, ReadOnlySequence<byte>> _tlsClientHelloBytesCallback;
+
+    public TlsListenerMiddleware(ConnectionDelegate next, Action<ConnectionContext, ReadOnlySequence<byte>> tlsClientHelloBytesCallback)
+    {
+        _next = next;
+        _tlsClientHelloBytesCallback = tlsClientHelloBytesCallback;
+    }
+
+    /// <summary>
+    /// Sniffs the TLS Client Hello message, and invokes a callback if found.
+    /// </summary>
+    internal async Task OnTlsClientHelloAsync(ConnectionContext connection)
+    {
+        var input = connection.Transport.Input;
+
+        while (true)
+        {
+            var result = await input.ReadAsync();
+            var buffer = result.Buffer;
+
+            // If the buffer length is less than 6 bytes (handshake + version + length + client-hello byte)
+            // and no more data is coming, we can't block in a loop here because we will not get more data
+            if (buffer.Length < 6 && result.IsCompleted)
+            {
+                break;
+            }
+
+            var parseState = TryParseClientHello(buffer, out var clientHelloBytes);
+
+            // no data is consumed, it will be processed by the follow-up middlewares
+            input.AdvanceTo(buffer.Start);
+
+            switch (parseState)
+            {
+                case ClientHelloParseState.NotEnoughData:
+                    continue;
+
+                case ClientHelloParseState.NotTlsClientHello:
+                    await _next(connection);
+                    return;
+
+                case ClientHelloParseState.ValidTlsClientHello:
+                    _tlsClientHelloBytesCallback(connection, clientHelloBytes);
+                    await _next(connection);
+                    return;
+            }
+        }
+
+        await _next(connection);
+    }
+
+    private static ClientHelloParseState TryParseClientHello(ReadOnlySequence<byte> buffer, out ReadOnlySequence<byte> clientHelloBytes)
+    {
+        clientHelloBytes = default;
+
+        if (buffer.Length < 6)
+        {
+            return ClientHelloParseState.NotEnoughData;
+        }
+
+        var reader = new SequenceReader<byte>(buffer);
+
+        // Content type must be 0x16 for TLS Handshake
+        if (!reader.TryRead(out byte contentType) || contentType != 0x16)
+        {
+            return ClientHelloParseState.NotTlsClientHello;
+        }
+
+        // Protocol version
+        if (!reader.TryReadBigEndian(out short version) || IsValidProtocolVersion(version) == false)
```

Comment:
```suggestion
        if (!reader.TryReadBigEndian(out short version) || !IsValidProtocolVersion(version))
```
or if your intention is readability (because the `!Is` may be hard to spot):
```suggestion
        if (!reader.TryReadBigEndian(out short version) || IsValidProtocolVersion(version) is false)
```
then the compiler will re-write it, but `MyBoolReturningFunction() == false` looks just wrong :wink:

Reply:
rewritten to `!IsValid...`, but I dont how "compiler will re-write it": [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+B6DACAMgSzBgDsBnGAE2wBcIaALGbAOgDkBRAFWwDEIBXYhQCG1fBGLZBFGFGwSm0bAFtoTYQHNYMZSWqlmAWABQWbJ0YsO3PtNHjJAG0IlypBvncAzfI6a1sAE8BKSFZBiYAWQBJbmciMhgjYxMAAQAGbFSARmYAIX4vL1lSAG4TE3xialliYUdscnrKLIAmc0dSAlIa4llI/AoKPwB3YVgTAG8TbFnsAAcofAA3USYcpCyUcyhAgAVx8gBhZz0ACRhHRwgACgAlGGEKAHliR0CAZRgAR34SIgAPMBAjUAHzYYCFYpQACUMzm02Mc2R2FWclgT3CAF5sH0Rtgvr9/jAHpioECQTBQTdIUVZDDyikkSjZvgvNgbgBCDEyKDMTi7UkUPL4DTsIT4YTEG4CaiNejQOXLEoObIw7BoNDYTnRUgANXqgz2UAgtEgjj1KokN2VUFIqphcOZLMRLNmAF94SyvSi2RzuY9efzBYGRWKJVKZfw5aQFVAlVbiK11ZrsLqDc4KMbTRBzZa7Q4bYnk9gsTivPVyE63dhXW7Pc6UT7kX6uTzZMHAkKw+KKJLpbL5YrUYmAMwprXpw1Zk1miAWxNFgsScfYTzYCudGDVt11lkNms+g9zH2LFZrLLZTbACDztP66fZucL5fS2PD232iQ7llY8Gfhw13cdIEHSMD2mwMwPg+XBsFaZhMhuECwPSZNmxrFklGQ0cwNmKCYOwUcEI5bCwJ/DC3Sw0CcOyPCcE4XAPmwXJENItV0IouYqPSHCILMBimNyWikOo1DyM45FuJw0c6PMRjmOYdoRJ4njxIk2YpPSFBSkg+j5NyGTlJwlAnQbIA=)

do you mind elaborating on how compiler will optimize it?

Reply:
I didn't mean the compiler will optimize this, rather it will rewrite it. So `boolean == false` or `boolean is false` to `!boolean`.

Check [this sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+B6DACAMgSzBgDsBnGAE2wBcIaALGbAOgDkBRAFWwDEIBXYhQCG1fBGLZBFGFGwSm0bAFtoTYQHNYMZSWqlmAWABQWbJ0YsO3PtNHjJAG0IlypBvncAzfI6a1sAE8BKSFZBiYAWQBJbmciMhgjYxMAAQAGbFSARmYAIX4vL1lSAG4TE3xialliYUdscnrKLIAmc0dSAlIa4llI/AoKPwB3YVgTAG8TbFnsAAcofAA3USYcpGwq6nMoQIAFcfIAYWc9AAkYR0cIAAoAJRhhCgB5YkdAgGUYAEd+EiIAB5gIEagA+bDAQrFKAAShmc2mxjmKOwqzksGe4QAvNg+iNsN8/gCYI8sVBgaCYGDblCirJYeUUsjUbN8F5sLcAISYmRQZicPZkih5fAadhCfDCYi3AQ7Uj0aA7ZYlBzZWHYNBobBc6KkABq9UG+ygEFokEc+tVEluKqgpDVsPhLNZSNZrNSAHZsNkme7sABfBGs4Oo9mcnlPPkCoVR0XiyXS2X8eWKqDK63EVoarXYPWG5wUE1miAWq32hy2zPZ7DY3FeerkZ3+t3+uZe7CtP3uoMu1GhlHh7m82QxwLC+MSihSmVyxppjMViQAZhz2vzRqLpvNEEtmarS+Iq623kbMGb7tbbay3uX3dZvf9A/b3rgvtDj7mocWKzWWWymzABAu55gam7Fjue6HrcCpKmimYXqy2IQnaDoSCe2DpAg6Q4e02BmJ8ny4J2zCZLcWE4ek2bPtecxKBRy44bMBFEdgy6kZyDE4YhtGovR2GMdkzE4JwuCfD6HHkQJ6TqjRvH8ekjF4WYonibkQlSYpVE8bxdFyFxy7CeYYkSe0mmMaucm0QpjEoKU+EiSZuSGeZ6QoM6vZAA=) (basically yours, but it returns a value, so you can see what Roslyn does a bit better).

Reply:
got it, thanks for explanation. I have changed to `!IsValid...`, so resolving the comment